[Encoding] Implement matmul_k encoding propagation across reshapes.

hanhanW opened the pull request:
The revision ports the sdxl propagation effort to the main branch. Ideally, we should implement it using interfaces and data-flow analysis.

It is a first step of the propagation, and we will incrementally enhance the encoding propagation pass.

Co-authored-by: MaheshRavishankar [mahesh.ravishankar@gmail.com](mailto:mahesh.ravishankar@gmail.com)

pashu123 commented on compiler/src/iree/compiler/DispatchCreation/PropagateEncodings.cpp:
```diff
@@ -0,0 +1,161 @@
+/// For a result of a `tensor.collapse_shape` given the indexing map
+/// in a consumer, return the iteration space expansion needed
+/// to swap the consumer and the `tensor.collapse_shape`.
+static std::tuple<unsigned, SmallVector<ReassociationIndices>>
```
```suggestion
static std::pair<unsigned, SmallVector<ReassociationIndices>>
```

pashu123 reviewed (commented):
LGTM!

Max191 reviewed (changes requested):
This propagation seems to be losing information. It works for the padding encoding case when we don't collapse the K dimension, but this would break with current data tiling encodings.

The reason it breaks is that we are overwriting the original indexing maps of the encoding, so the access pattern of the original matmul is lost. It will already cause things to break in the data tiling encodings because we don't support materialization for multiple M/N/K dimensions.

hanhanW committed:
[Encoding] Implement basic encoding propagation pass across reshapes.

Note: it only works for matmul_k encoding atm, we should invest the
interface way.

Signed-off-by: hanhanW <hanhan0912@gmail.com>

hanhanW committed:
Tighten the propagation condition

Signed-off-by: hanhanW <hanhan0912@gmail.com>

Max191 commented on compiler/src/iree/compiler/DispatchCreation/PropagateEncodings.cpp:
```diff
@@ -0,0 +1,122 @@
+    return rewriter.notifyMatchFailure(encodingOp,
+                                       "expected a collapse_shape producer");
+  }
+  if (!IREE::Flow::isNonNullAndOutsideDispatch(encodingOp) ||
+      !IREE::Flow::isNonNullAndOutsideDispatch(collapseOp)) {
+    return rewriter.notifyMatchFailure(
+        encodingOp, "expected that both operations are outside dispatch");
```
Do we need them to be outside of a dispatch? I think the requirement could be that the collapseOp is not outside of a dispatch when the encodingOp is inside of a dispatch (because swapping them could cause new hoisting/fusion). If they are both inside of a dispatch, then the propagation could still happen.

(I don't have a strong opinion on this, mostly just pointing it out. I don't think the propagation when both are inside of the dispatch really helps us much anyway)

hanhanW replied:
> If they are both inside of a dispatch, then the propagation could still happen.

We'll use interfaces anyway, let's iterate on it when we implement the method for dispatch region ops and other ops. At least this is what we've seen in llama, so starting from here sounds okay to me.

The PR has been floating for a long time, so I really want to land some basic first.

Max191 replied:
Sounds good!

Max191 commented on compiler/src/iree/compiler/DispatchCreation/PropagateEncodings.cpp:
```diff
@@ -0,0 +1,122 @@
+          llvm::seq<int32_t>(numDims, numDims + reassociation.size())));
+    }
+    numDims += reassociation.size();
+  }
```
nit: I don't think the numDims is necessary. It could just be:
```suggestion
  SmallVector<int32_t> newKDims;
  for (int32_t kDim : dDims) {
    newKDims.append(reassociationMaps[kDim]);
  }
```

Max191 replied:
It could also be combined with the above verification loop if doing it this way.

hanhanW replied:
Good point, thanks!

Max191 commented on compiler/src/iree/compiler/DispatchCreation/PropagateEncodings.cpp:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  // Get a mapping from original iteration space to expanded iteration space.
+  MLIRContext *ctx = rewriter.getContext();
```
nit: move `ctx` to below where it is used.

Max191 commented on compiler/src/iree/compiler/DispatchCreation/test/propagate_encodings.mlir:
```diff
@@ -0,0 +1,33 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(util.func(iree-dispatch-creation-propagate-encodings))" --split-input-file %s | FileCheck %s
```
Can we add a test that shows hoisting does not happen when k dims are collapsed?

hanhanW replied:
What do you mean? This pass does not know anything about hoisting?

Max191 replied:
Sorry, I mean propagation. Getting things mixed up today :p

hanhanW replied:
I think we have the test case, see the second test. The set_encoding is in the between of two collapse shape ops.

Max191 reviewed (commented):
Overall looks good! Just some nits

hanhanW committed:
Merge branch 'main' into basic-encoding-propagation

hanhanW committed:
address comments

Signed-off-by: hanhanW <hanhan0912@gmail.com>

Max191 reviewed (approved):
A test that shows propagation being blocked by collapsing k_dims would be good to have, and then LGTM!
